File: swift/BCComponents/Sources/BCComponents/XID.swift
import BCUR
import BCTags
import DCBOR
import Foundation

public protocol XIDProvider {
    func xid() -> XID
}

public struct XID: Equatable, Hashable, Comparable, Sendable {
    public static let xidSize = 32

    private let value: Data

    public init(_ value: Data) throws(BCComponentsError) {
        try requireLength(value, expected: Self.xidSize, name: "XID")
        self.value = value
    }

    public static func fromData(_ value: Data) throws(BCComponentsError) -> XID {
        try XID(value)
    }

    public static func fromHex(_ hex: String) throws(BCComponentsError) -> XID {
        try XID(parseHex(hex))
    }

    public static func new(genesisKey: some CBOREncodable) -> XID {
        let digest = Digest.fromImage(genesisKey.cborData)
        return try! XID(digest.data)
    }

    public func validate(genesisKey: some CBOREncodable) -> Bool {
        let digest = Digest.fromImage(genesisKey.cborData)
        return digest.data == value
    }

    public var data: Data {
        value
    }

    public func asBytes() -> Data {
        value
    }

    public func toHex() -> String {
        hexEncode(value)
    }

    public func shortDescription() -> String {
        refHexShort()
    }

    public func bytewordsIdentifier(prefix: Bool) -> String {
        refBytewords(prefix ? "🅧" : nil)
    }

    public func bytemojiIdentifier(prefix: Bool) -> String {
        refBytemoji(prefix ? "🅧" : nil)
    }

    public static func < (lhs: XID, rhs: XID) -> Bool {
        lhs.value.lexicographicallyPrecedes(rhs.value)
    }
}

extension XID: XIDProvider {
    public func xid() -> XID {
        self
    }
}

extension XID: ReferenceProvider {
    public func reference() -> Reference {
        try! Reference(value)
    }
}

extension XID: CustomStringConvertible {
    public var description: String {
        "XID(\(shortDescription()))"
    }
}

extension XID: CustomDebugStringConvertible {
    public var debugDescription: String {
        "XID(\(toHex()))"
    }
}

extension XID: CBORTaggedEncodable {
    public static var cborTags: [Tag] {
        [.xid]
    }

    public var untaggedCBOR: CBOR {
        .bytes(value)
    }
}

extension XID: CBORTaggedDecodable {
    public init(untaggedCBOR: CBOR) throws {
        try self.init(byteString(untaggedCBOR))
    }
}

extension XID: URCodable {}

File: typescript/bc-components/src/id/xid.ts
import { type Cbor, createTag, toByteString } from '@bc/dcbor';
import { TAG_XID } from '@bc/tags';

import type { Tag } from '@bc/dcbor';

import { identifier as bytewordsId, bytemojiIdentifier as bytemojiId } from '@bc/ur';

import { Digest } from '../digest.js';
import { Reference } from '../reference.js';
import { decodeTaggedCbor, defaultTaggedCbor, defaultTaggedCborData, fromTaggedUrString, toUrString } from '../cbor-ur.js';
import { bytesEqual, hexDecode, hexEncode, requireLength, toHexShort } from '../utils.js';

const XID_TAG: Tag = createTag(TAG_XID, 'xid');

export interface XIDProvider {
    xid(): XID;
}

export interface SigningPublicKeyLike {
    taggedCborData(): Uint8Array;
}

export class XID {
    static readonly XID_SIZE = 32;

    readonly #data: Uint8Array;

    constructor(data: Uint8Array) {
        this.#data = requireLength(data, XID.XID_SIZE, 'XID');
    }

    static fromData(data: Uint8Array): XID {
        return new XID(new Uint8Array(data));
    }

    static fromDataRef(data: Uint8Array): XID {
        return XID.fromData(data);
    }

    static new(genesisKey: SigningPublicKeyLike): XID {
        return XID.fromData(Digest.fromImage(genesisKey.taggedCborData()).data);
    }

    validate(genesisKey: SigningPublicKeyLike): boolean {
        return bytesEqual(this.#data, XID.new(genesisKey).#data);
    }

    static fromHex(hex: string): XID {
        return XID.fromData(hexDecode(hex));
    }

    toHex(): string {
        return hexEncode(this.#data);
    }

    get shortDescription(): string {
        return toHexShort(this.#data);
    }

    bytewordsIdentifier(prefix: boolean): string {
        const core = bytewordsId(this.#data.slice(0, 4)).toUpperCase();
        return prefix ? `🅧 ${core}` : core;
    }

    bytemojiIdentifier(prefix: boolean): string {
        const core = bytemojiId(this.#data.slice(0, 4));
        return prefix ? `🅧 ${core}` : core;
    }

    get data(): Uint8Array {
        return new Uint8Array(this.#data);
    }

    asBytes(): Uint8Array {
        return this.data;
    }

    reference(): Reference {
        return Reference.fromData(this.#data);
    }

    xid(): XID {
        return this;
    }

    equals(other: unknown): boolean {
        return other instanceof XID && bytesEqual(this.#data, other.#data);
    }

    cborTags(): Tag[] {
        return [XID_TAG];
    }

    untaggedCbor(): Cbor {
        return toByteString(this.#data);
    }

    taggedCbor(): Cbor {
        return defaultTaggedCbor(this);
    }

    taggedCborData(): Uint8Array {
        return defaultTaggedCborData(this);
    }

    urString(): string {
        return toUrString(this);
    }

    static fromURString(value: string): XID {
        return fromTaggedUrString(XID, value);
    }

    static cborTags(): Tag[] {
        return [XID_TAG];
    }

    static fromUntaggedCbor(cbor: Cbor): XID {
        return XID.fromData(cbor.toByteString());
    }

    static fromCbor(cbor: Cbor): XID {
        return decodeTaggedCbor(XID, cbor);
    }

    toString(): string {
        return `XID(${this.shortDescription})`;
    }
}
